core/src/main/java/org/infinispan/xsite/BaseBackupReceiver.java

```diff
@@ -72,71 +88,202 @@ public final Cache getCache() {
    }
 
    @Override
-   public final Object handleRemoteCommand(VisitableCommand command) throws Throwable {
-      return command.acceptVisitor(null, siteUpdater);
+   public final CompletionStage<Void> handleRemoteCommand(VisitableCommand command, boolean preserveOrder) {
+      try {
+         DefaultHandler visitor = preserveOrder ? orderedHandler : defaultHandler;
+         //noinspection unchecked
+         return (CompletableFuture<Void>) command.acceptVisitor(null, visitor);
+      } catch (Throwable throwable) {
+         return completedExceptionFuture(throwable);
+      }
    }
 
-   protected final void assertAllowInvocation() {
+
+   final <T> CompletableFuture<T> checkInvocationAllowedFuture() {
       ComponentStatus status = cache.getStatus();
       if (!status.allowInvocations()) {
-         throw new CacheException("Cache is stopping or terminated: " + status);
+         return completedExceptionFuture(new CacheException("Cache is stopping or terminated: " + status));
+      }
+      return null;
+   }
+
+   final Void assertAllowInvocationFunction(Object ignoredRetVal) {
+      //the put operation can fail silently. check in the end and it is better to resend the chunk than to lose keys.
+      ComponentStatus status = cache.getStatus();
+      if (!status.allowInvocations()) {
+         throw asCompletionException(new CacheException("Cache is stopping or terminated: " + status));
+      }
+      return null;
+   }
+
+   private static final class OrderedHandler extends DefaultHandler {
```

Comment:
> `DefaultHandler` should be first.
what do you mean?

> But I don't really like `OrderedHandler` extending `DefaultHandler`, because it's not obvious why some methods are overridden and call `super`, `visitPrepareCommand` is overridden but doesn't call `super`, and `visitCommitCommand` and `visitRollbackCommand` are not overridden at all. I'd rather have them as independent classes and the common code in outside methods.

`visitPrepareCommand` doesn't call support because I'm using `sequencer.orderBlockingTask()` to avoid creating a new `CompletableFuture` (that is created if I invoke  `super`). If I drop `orderBlockingTask()`, then I have to use `super`.

`visitCommitCommand` and `visitRollbackCommand` doesn't have a keys to be ordered so why override? Also, 2PC commit with async-xsite won't work so you will never have rollbacks or commits here.

Reply:
> > `DefaultHandler` should be first.
> 
> what do you mean?

I meant the order of the classes, it's clearer if the base class is first.

> `visitPrepareCommand` doesn't call support because I'm using `sequencer.orderBlockingTask()` to avoid creating a new `CompletableFuture` (that is created if I invoke `super`). If I drop `orderBlockingTask()`, then I have to use `super`.

Yeah, looks better now :)

> `visitCommitCommand` and `visitRollbackCommand` doesn't have a keys to be ordered so why override? Also, 2PC commit with async-xsite won't work so you will never have rollbacks or commits here.

Sure, you know the reason, but it's not obvious for the reader (i.e. me), because the class is called `OrderedHandler`, not `AsyncBackupHandler` :)

So maybe you should override `visitCommitCommand` and `visitRollbackCommand` to throw an exception, documenting this way that we only expect one-phase prepare commands?